### DISCUSSION TITLE
Why does Props.Create use Activator even when passing an expression?

### DISCUSSION THREAD

**Discussion by chylex** [akkadotnet/akka.net#7104] [Q&A] [answered]
The `Props Create<TActor>(Expression<Func<TActor>> factory)` overload takes an `Expression`, which if it is a `NewExpression`, creates `Props` with an `ActivatorProducer`. When I use the recommended static factory method, it makes sense to me (and Rider recommends this) to make the actor constructor private. However, this crashes at runtime, because `Activator` cannot call private constructors.

This is really confusing to me, so I'm wondering what's the rationale?

Normally in .NET, if an `Expression` calls a private constructor, compiling it and calling it elsewhere works - except in Akka.NET - making Akka.NET's behavior counter-intuitive.

It's also not clear why use an `Expression` at all, as opposed to a `Func`. The `Props` class already includes a `FactoryConsumer` implementation that works with a `Func`, but as far as I can tell, `FactoryConsumer` is not used anywhere.

I might try copying `FactoryConsumer` and using it via `Props.CreateBy(IIndirectActorProducer)`, so I can make actor constructors private and avoid `Activator`, but I guess there's a good reason why Akka.NET doesn't work this way by default?

**Comment by to11mtm** [comment, marked as answer]
Do you happen to have a repro? I can look into this more when back from
Vacation to see if we can enhance the expression handling in some way...

As far as the reason for an expression rather than a function, my general
assumption is that because Props are intended to be serializable, an
expression is typically easier to decompose into a constructor and args.


That said, we may be able to add some form of static factory for the parser
(we already have some custom parse logic to try and avoid .Compile()) but
an example case will make it easier per above.

**Reply by Aaronontheweb** [reply]
> As far as the reason for an expression rather than a function, my general
assumption is that because Props are intended to be serializable, an
expression is typically easier to decompose into a constructor and args.

That is correct

**Reply by chylex** [reply]
Minimal repro:

```csharp
public class Program {
	public static void Main() {
		using var actorSystem = ActorSystem.Create("Test");
		actorSystem.ActorOf(Actor.Factory(5));
	}
}

sealed class Actor : ReceiveActor {
	public static Props Factory(int value) {
		return Props.Create(() => new Actor(value));
	}

	private int Value { get; }

	private Actor(int value) {
		Value = value;
	}
}

```

It should be basically the same as https://getakka.net/articles/actors/receive-actor-api.html#recommended-practices, except with a private constructor. With the private constructor, Akka.NET throws at runtime:

```
[ERROR][02.26.2024 21:44:08.850Z][Thread 0022][akka://Test/user/$a]
Error while creating actor instance of type Controller.Actor with 1 args: (5)
Cause: [akka://Test/user/$a#467941579]: Akka.Actor.ActorInitializationException: Exception during creation                                               
 ---> System.TypeLoadException: Error while creating actor instance of type Controller.Actor with 1 args: (5)   
 ---> System.MissingMethodException: Constructor on type 'Controller.Actor' not found.                          
   at System.RuntimeType.CreateInstanceImpl(BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture)
   at System.Activator.CreateInstance(Type type, Object[] args)                                                         
   at Akka.Actor.Props.ActivatorProducer.Produce()                                                                      
   at Akka.Actor.Props.NewActor()                                                                                       
   --- End of inner exception stack trace ---                                                                           
   at Akka.Actor.Props.NewActor()                                                                                       
   at Akka.Actor.ActorCell.CreateNewActorInstance()                                                                     
   at Akka.Actor.ActorCell.<>c__DisplayClass116_0.<NewActor>b__0()                                                      
   at Akka.Actor.ActorCell.UseThreadContext(Action action)                                                              
   at Akka.Actor.ActorCell.NewActor()                                                                                   
   at Akka.Actor.ActorCell.Create(Exception failure)                                                                    
   --- End of inner exception stack trace ---                                                                           
   at Akka.Actor.ActorCell.Create(Exception failure)                                                                    
   at Akka.Actor.ActorCell.SysMsgInvokeAll(EarliestFirstSystemMessageList messages, Int32 currentState)    
```

**Reply by chylex** [reply]
Personally I don't need serialization, so making a custom `IIndirectActorProducer` is good enough for me. This answers my question, but should I mark it as the answer now, or keep it open until you've had a chance to look at the repro example?

**Reply by Aaronontheweb** [reply]
Glad you were able to find the `IIndirectActorProducer` - that's what I was going to suggest! I don't think we'll be making any significant changes to `Props` (as far as `NewExpression` usage goes) any time soon since that'd break features like remote actor deployment, but I'll mark your solution as the answer here in case other users run into this.

**Reply by chylex** [reply]
Cool; I already mentioned `IIndirectActorProducer` indirectly (heh) in my original post, but I'll edit to make it clearer and mark [to11mtm](https://github.com/to11mtm)'s response as the answer since that directly answers the "why" in my original question.

Thanks to both for answering so quickly!

**Reply by chylex** [reply]
I have opened an improvement for Rider (https://youtrack.jetbrains.com/issue/RIDER-106480) to stop suggesting making actor constructors private. If you find a way to allow private constructors, please let me know so I can update the issue.
